compiler/resolution/functionResolution.cpp

```diff
@@ -11255,6 +11255,82 @@ static void checkNoVoidFields()
   }
 }
 
+/************************************* | **************************************
+*                                                                             *
+*  Find specially named methods such as 'hash' that are placed on types that  *
+*  don't have the corresponding interface. This is executed here because      *
+*  resolution cleans up unused functions and interface statements, but we want*
+*  to warn even if a function isn't used.                                     *
+*                                                                             *
+************************************** | *************************************/
+
+struct SpeciallyNamedMethodInfo {
+  std::map<std::string, FnSymbol*> speciallyNamedMethods;
+};
+
+using SpeciallyNamedMethodKey = std::pair<AggregateType*, InterfaceSymbol*>;
+using SpecialMethodMap = std::map<SpeciallyNamedMethodKey, SpeciallyNamedMethodInfo>;
+
+static void checkSpeciallyNamedMethods() {
+  static const std::unordered_map<std::string, InterfaceSymbol*> reservedNames = {
+    { "hash", gHashable },
+  };
+
+  SpecialMethodMap flagged;
+
+  // TODO: for now, this will simply not warn for classes, because all classe
+  // implement hashable, and because overriding hash should be allowed. When
+  // other special methods are converted, this logic will need to be updated
+  // to handle them.
+
+  for_alive_in_Vec(FnSymbol, fn, gFnSymbols) {
+    if (!fn->isMethod()) continue;
+    if (fn->isCompilerGenerated() || fn->hasFlag(FLAG_FIELD_ACCESSOR)) continue;
+
+    auto reservedIter = reservedNames.find(fn->name);
+    if (reservedIter == reservedNames.end()) continue;
+    auto receiverType = fn->getReceiverType();
+
+    auto at = toAggregateType(receiverType);
+    if (!at || !at->isRecord()) continue;
```

Comment:
Are we only warning about hash for records, not for classes?  Should this eliminate unions instead?

Reply:
All classes implement the `hashable` interface by virtue of the parent class implementing the hashable interface (and having a `proc hash`). This warning would be spurious under the current TODO because the logic here does not handle a method `hash` defined on a child class while a parent class `implements hashable`.